File: python-glue/unified_ai/resilience/circuit_breaker.py
"""Circuit breaker pattern implementation"""

import asyncio
import time
from enum import Enum
from typing import TypeVar, Callable, Awaitable, Optional
from dataclasses import dataclass, field
from threading import Lock

T = TypeVar("T")


class CircuitState(Enum):
    CLOSED = "closed"
    OPEN = "open"
    HALF_OPEN = "half_open"


@dataclass
class CircuitBreaker:
    """Circuit breaker for preventing cascading failures"""
    
    name: str
    failure_threshold: int = 5
    success_threshold: int = 2
    timeout: float = 60.0
    
    _state: CircuitState = field(default=CircuitState.CLOSED, init=False)
    _failures: int = field(default=0, init=False)
    _successes: int = field(default=0, init=False)
    _last_failure_time: Optional[float] = field(default=None, init=False)
    _lock: Lock = field(default_factory=Lock, init=False)
    
    @property
    def state(self) -> CircuitState:
        """Get current circuit breaker state"""
        with self._lock:
            return self._state
    
    def _check_state(self) -> None:
        """Check and update circuit breaker state"""
        with self._lock:
            if self._state == CircuitState.OPEN:
                if self._last_failure_time:
                    elapsed = time.time() - self._last_failure_time
                    if elapsed >= self.timeout:
                        self._state = CircuitState.HALF_OPEN
                        self._successes = 0
    
    def _on_success(self) -> None:
        """Handle successful call"""
        with self._lock:
            self._failures = 0
            if self._state == CircuitState.HALF_OPEN:
                self._successes += 1
                if self._successes >= self.success_threshold:
                    self._state = CircuitState.CLOSED
                    self._successes = 0
    
    def _on_failure(self) -> None:
        """Handle failed call"""
        with self._lock:
            self._failures += 1
            self._last_failure_time = time.time()
            
            if self._state == CircuitState.HALF_OPEN:
                self._state = CircuitState.OPEN
            elif self._state == CircuitState.CLOSED:
                if self._failures >= self.failure_threshold:
                    self._state = CircuitState.OPEN
    
    async def call(self, func: Callable[..., Awaitable[T]], *args, **kwargs) -> T:
        """Execute function with circuit breaker protection"""
        self._check_state()
        
        if self._state == CircuitState.OPEN:
            raise RuntimeError(
                f"Circuit breaker '{self.name}' is OPEN. "
                f"Retry after {self.timeout}s"
            )
        
        try:
            result = await func(*args, **kwargs)
            self._on_success()
            return result
        except Exception as e:
            self._on_failure()
            raise

File: rust-core/src/resilience/circuit_breaker.rs
use crate::error::{OrchestratorError, Result};
use std::sync::{Arc, Mutex};
use std::time::{Duration, Instant};

#[derive(Debug, Clone, Copy, PartialEq)]
pub enum CircuitState {
    Closed,
    Open,
    HalfOpen,
}

#[derive(Debug)]
struct CircuitBreakerInner {
    failure_threshold: u32,
    success_threshold: u32,
    timeout: Duration,
    state: CircuitState,
    failures: u32,
    successes: u32,
    last_failure_time: Option<Instant>,
}

impl CircuitBreakerInner {
    fn new(failure_threshold: u32, timeout: Duration) -> Self {
        Self {
            failure_threshold,
            success_threshold: 2,
            timeout,
            state: CircuitState::Closed,
            failures: 0,
            successes: 0,
            last_failure_time: None,
        }
    }
    
    fn on_success(&mut self) {
        self.failures = 0;
        match self.state {
            CircuitState::HalfOpen => {
                self.successes += 1;
                if self.successes >= self.success_threshold {
                    self.state = CircuitState::Closed;
                    self.successes = 0;
                }
            }
            _ => {}
        }
    }
    
    fn on_failure(&mut self) {
        self.failures += 1;
        self.last_failure_time = Some(Instant::now());
        
        match self.state {
            CircuitState::HalfOpen => {
                self.state = CircuitState::Open;
            }
            CircuitState::Closed => {
                if self.failures >= self.failure_threshold {
                    self.state = CircuitState::Open;
                }
            }
            _ => {}
        }
    }
    
    fn check_state(&mut self) -> Result<()> {
        match self.state {
            CircuitState::Open => {
                if let Some(last_failure) = self.last_failure_time {
                    if last_failure.elapsed() >= self.timeout {
                        self.state = CircuitState::HalfOpen;
                        self.successes = 0;
                        Ok(())
                    } else {
                        Err(OrchestratorError::CircuitBreakerOpen(
                            format!("Circuit breaker is open. Retry after {:?}", 
                                self.timeout - last_failure.elapsed())
                        ))
                    }
                } else {
                    Err(OrchestratorError::CircuitBreakerOpen(
                        "Circuit breaker is open".to_string()
                    ))
                }
            }
            CircuitState::HalfOpen | CircuitState::Closed => Ok(()),
        }
    }
}

#[derive(Debug, Clone)]
pub struct CircuitBreaker {
    inner: Arc<Mutex<CircuitBreakerInner>>,
    name: String,
}

impl CircuitBreaker {
    pub fn new(name: impl Into<String>, failure_threshold: u32, timeout: Duration) -> Self {
        Self {
            inner: Arc::new(Mutex::new(CircuitBreakerInner::new(failure_threshold, timeout))),
            name: name.into(),
        }
    }
    
    pub fn state(&self) -> CircuitState {
        self.inner.lock().unwrap().state
    }
    
    pub async fn call<T, F, Fut>(&self, f: F) -> Result<T>
    where
        F: FnOnce() -> Fut,
        Fut: std::future::Future<Output = Result<T>>,
    {
        // Check if we can proceed
        {
            let mut inner = self.inner.lock().unwrap();
            inner.check_state()?;
        }
        
        match f().await {
            Ok(result) => {
                let mut inner = self.inner.lock().unwrap();
                inner.on_success();
                Ok(result)
            }
            Err(e) => {
                let mut inner = self.inner.lock().unwrap();
                inner.on_failure();
                Err(e)
            }
        }
    }
}
